0: succesful
2: shutdown due to interrupted exception in thread
